Framework/MDAlgorithms/src/Integrate3DEvents.cpp

```diff
@@ -69,6 +72,266 @@ void Integrate3DEvents::addEvents(
+  const auto core = numInEllipsoid(events, eigen_vectors, sigmas);
+  const auto peak = numInEllipsoid(events, eigen_vectors, peakRadii);
+  const auto ratio = pow(r1, 3) / (pow(r3, 3) - pow(r2, 3));
+
+  inti = peak - ratio * backgrd;
+  sigi = sqrt(peak + ratio * ratio * backgrd);
+
+  if (inti < 0) {
+    inti = 0;
+    sigi = 0;
+  }
+
+  // compute the fraction of peak within the standard core
+  const auto total = (core + peak) - ratio * backgrd;
+  const auto frac = std::min(1.0, std::abs(inti / total));
+  // compute the uncertainty in the fraction
+  const auto df_ds_core = (1 - frac) / peak;
+  const auto df_ds_peak = frac / peak;
+  const auto fracError =
+      sqrt(peak * pow(df_ds_core, 2) + core * pow(df_ds_peak, 2));
+
+  // create the peaks shape for the strong peak
+  const auto shape = boost::make_shared<const PeakShapeEllipsoid>(
+      eigen_vectors, peakRadii, abcBackgroundInnerRadii,
+      abcBackgroundOuterRadii, Mantid::Kernel::QLab,
+      "IntegrateEllipsoidsTwoStep");
+
+  return std::make_pair(shape, std::make_tuple(frac, fracError, max_sigma));
+}
+
+boost::shared_ptr<const Geometry::PeakShape>
+Integrate3DEvents::integrateWeakPeak(
+    const IntegrationParameters &params, PeakShapeEllipsoid_const_sptr shape,
+    const std::tuple<double, double, double> &libPeak, const V3D &center,
+    double &inti, double &sigi) {
+
+  inti = 0.0; // default values, in case something
+  sigi = 0.0; // is wrong with the peak.
+
+  auto result = getEvents(center);
+  if (!result)
+    return boost::make_shared<NoShape>();
+
+  const auto &events = result.get();
+
+  const auto &directions = shape->directions();
+  const auto &abcBackgroundInnerRadii = shape->abcRadiiBackgroundInner();
+  const auto &abcBackgroundOuterRadii = shape->abcRadiiBackgroundOuter();
+  const auto &abcRadii = shape->abcRadii();
+
+  const auto max_sigma = std::get<2>(libPeak);
+  auto rValues = calculateRadiusFactors(params, max_sigma);
+
+  const auto isPeakOnDetector =
+      correctForDetectorEdges(rValues, params.E1Vectors, center, abcRadii,
+                              abcBackgroundInnerRadii, abcBackgroundOuterRadii);
+
+  if (!isPeakOnDetector)
+    return shape;
+
+  const double r1 = std::get<0>(rValues), r2 = std::get<1>(rValues),
+               r3 = std::get<2>(rValues);
+
+  // integrate
+  double backgrd = numInEllipsoidBkg(
+      events, directions, abcBackgroundOuterRadii, abcBackgroundInnerRadii);
+  double peak_w_back = numInEllipsoid(events, directions, abcRadii);
+  double ratio = pow(r1, 3) / (pow(r3, 3) - pow(r2, 3));
+
+  const auto frac = std::get<0>(libPeak);
+  const auto fracError = std::get<1>(libPeak);
+
+  inti = peak_w_back - ratio * backgrd;
+  sigi = inti + ratio * ratio * backgrd;
+
+  // correct for fractional intensity
+  sigi = sigi / pow(inti, 2);
+  sigi += pow((fracError / frac), 2);
+
+  inti = inti * frac;
+  sigi = sqrt(sigi) * inti;
```

Comment:
It's error in the integration estimate. So I think it should be the same units as the integrated value? Not 100% sure about that though.